scripts/airgapped/get-all-images.py

```diff
@@ -0,0 +1,177 @@
+        return True
+
+    return False
+
+
+def bundle_app_contains_gh_metadata(app: dict) -> bool:
+    """
+    Given an application in a bundle check if it contains github metadata keys
+    to be able to be parsed properly.
+    """
+    if is_dependency_app(app):
+        return True
+
+    if GH_REPO_KEY in app and GH_BRANCH_KEY in app:
+        return True
+
+    return False
+
+
+def clone_repo(repo_name: str, branch: str) -> str:
+    """Clones locally a repo and returns the folder created."""
+    repo_url = f"https://github.com/canonical/{repo_name}.git"
+
+    logging.info(f"Cloning repo {repo_url}")
+    repo = git.Repo.clone_from(repo_url, str(repo_name))
+
+    logging.info(f"Checking out to branch {branch}")
+    repo.git.checkout(branch)
+
+    return repo_name
+
+
+def get_analytics_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a charm developed by us, by:
+    1. Cloning the repo of the charm
+    2. Running the repo's tools/get-images.sh
+    3. Delete the repo
+
+    If the tools/get-images.sh of a repo fails for any reason then this
+    script will also fail.
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_REPO_KEY], app[GH_BRANCH_KEY])
+
+    logging.info(f"Executing repo's {GET_IMAGES_SH} script")
+    process = subprocess.Popen(["bash", "tools/get-images.sh"], cwd=repo_dir,
+                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = process.communicate()
+    if stderr:
+        raise ValueError("Script '%s' for charm '%s' had error logs: \n%s" %
+                         (GET_IMAGES_SH, app["charm"], stderr.decode("utf-8"))
+                         )
+    images = stdout.decode("utf-8").split("\n")
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_dependency_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a dependency charm by:
+    1. Cloning the repo of the charm
+    2. Looking at its metadata.yaml for "upstream-source" keys
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_bundle_images(bundle_dict: dict) -> list[str]:
+    """Return a list of images used by a bundle"""
+    images = []
+
+    for app_name, app in bundle_dict["applications"].items():
+        logging.info(f"Handling app {app_name}")
+
+        # exclude repos we know don't have images
+        if app_name in EXCLUDE_REPOS:
```

Comment:
I like the idea, although in that case it's about repos that we might want to include **or** exclude.

In case of `kubeflow-roles` we **always** want to exclude it since it doesn't contain any images, so in my mind that feature is orthogonal to always excluding some charms/repos.

Reply:
Perhaps I was not very clear in my suggestion. If we had the option of exluding repos, let's say `get-all-images.py --exclude=istio,mysql,mlflow`, you could append those repos to exclude to the `EXCLUDE_REPOS` list, so they will be ignored alongside `kubeflow-roles`. Alternatively, we can have a `get-all-images.py --repos=kfp-operators,istio` that gets the images only on the repos I want. This can make this script more flexible and we can use it for things like: "I want to know if all the charms in the kfp repo are using the rocks w/o me having to look into the scans action".

Again, not for blocking this PR, but to consider.

Reply:
Ah OK. Added a comment also in the code in case someone might need to go and extend the list to instead thing of having a `--exclude` argument (rather than keep extending the list)

462c1fe